pkg/operator/controller/ingress-routes/controller.go

```diff
@@ -0,0 +1,250 @@
+package ingress
+
+import (
+	"context"
+	"fmt"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	configv1 "github.com/openshift/api/config/v1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "ingress_routes_controller"
+)
+
+var (
+	log = logf.Logger.WithName(controllerName)
+)
+
+// New creates the ingress routes controller from configuration. This is the controller
+// that handles all the logic for generating roles and rolebindings for operators that
+// include routes with configurable hostnames and service certificate.
+func New(mgr manager.Manager, config Config) (controller.Controller, error) {
+	reconciler := &reconciler{
+		config: config,
+		client: mgr.GetClient(),
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	if err := c.Watch(&source.Kind{Type: &configv1.Ingress{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+	// TODO: The operator only watches for events in the openshift-ingress-operator namespace,
+	// changes to generated role and rolebindings in the openshift-config namespace are ignored.
+	// Find a way to trigger events outside of the openshift-ingress-operator namespace.
+	if err := c.Watch(&source.Kind{Type: &rbacv1.Role{}}, &handler.EnqueueRequestForOwner{OwnerType: &configv1.Ingress{}}); err != nil {
+		return nil, err
+	}
+	if err := c.Watch(&source.Kind{Type: &rbacv1.RoleBinding{}}, &handler.EnqueueRequestForOwner{OwnerType: &configv1.Ingress{}}); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Config holds all the things necessary for the controller to run.
+type Config struct {
+	SecretNamespace string
+}
+
+// reconciler handles the actual ingress reconciliation logic in response to
+// events.
+type reconciler struct {
+	config Config
+	client client.Client
+}
+
+// Reconcile expects request to refer to a ingress in the operator namespace,
+// and will do all the work to ensure the ingress is in the desired state.
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
```

Comment:
how do errors here show up as degraded in the operator.  I think this and the lbirary-go apply are the most critical to work out ot decide if it can go here or needs to go elsewhere.